Bug: TypeScript 4.6 report type errors on `trigger.response`

## Description

While upgrading TypeScript to 4.6, it starts to keep reporting type errors in `test/helper/test_trigger.ts`

```bash
$ npm run tsc-no-emit

> cactbot@0.24.24 tsc-no-emit
> tsc --noEmit

test/helper/test_trigger.ts:533:30 - error TS2349: This expression is not callable.
  Each member of the union type 'ResponseFunc<RaidbossData, Matches> | (ResponseFunc<RaidbossData, Matches & ToMatches<{ readonly type: 0; readonly timestamp: 1; readonly code: 2; readonly name: 3; readonly line: 4; }>> & ResponseFunc<...>) | ... 103 more ... | ({ ...; } & ResponseFunc<...>)' has signatures, but none of those signatures are compatible with each other.

533           response = trigger.response(data, {}, output) ?? {};
                                 ~~~~~~~~


Found 1 error in test/helper/test_trigger.ts:533
```

Alhtough there is an if-check on top that assert the `response` property should be a function, TypeScript didn't successfully narrow the type to function only:

https://github.com/quisquous/cactbot/blob/951f3350d8bd52d8712b6559c27f724d95c31c93/test/helper/test_trigger.ts#L521-L533

------

@valarnin Hi type wizard, FYI if you can help with the fixing.

--- a/package.json
+++ b/package.json
@@ -97,7 +97,7 @@
     "ts-node": "^10.7.0",
     "tsconfig-paths": "^3.9.0",
     "ttypescript": "^1.5.12",
-    "typescript": "^4.5.0",
+    "typescript": "^4.6.4",
     "webpack": "^5.64.3",
     "webpack-cli": "^4.9.1",
     "webpack-dev-server": "^4.5.0",

--- a/test/helper/test_trigger.ts
+++ b/test/helper/test_trigger.ts
@@ -24,6 +24,7 @@ import {
   LooseTriggerSet,
   Output,
   OutputStrings,
+  ResponseFunc,
   TriggerFunc,
 } from '../../types/trigger';
 
@@ -48,6 +49,10 @@ const netRegexLanguages: (keyof LooseTrigger)[] = [
 ];
 
 const isKeyInTriggerFunctions = (arr: string[], key: string): boolean => arr.includes(key);
+
+const isResponseFunc = (func: unknown): func is ResponseFunc<RaidbossData, Matches> => {
+  return typeof func === 'function';
+};
 
 const createTriggerRegexString = (str: string): RegExp => {
   return new RegExp(`(?:regex|triggerRegex)(?:|\\w{2}): \/${str}\/`, 'g');
@@ -527,14 +532,18 @@ const testTriggerFile = (file: string) => {
             );
             continue;
           }
-          const output = new TestOutputProxy(trigger, outputStrings) as Output;
-          // Call the function to get the outputStrings.
-          const data = (triggerSet.initData?.() ?? {}) as RaidbossData;
-          response = trigger.response(data, {}, output) ?? {};
 
-          if (typeof outputStrings !== 'object') {
-            assert.fail(`'${trigger.id}' built-in response did not set outputStrings.`);
-            continue;
+          const output = new TestOutputProxy(trigger, outputStrings) as Output;
+          const responseFunc = trigger.response;
+          if (isResponseFunc(responseFunc)) {
+            // Call the function to get the outputStrings.
+            const data = (triggerSet.initData?.() ?? {}) as RaidbossData;
+            response = responseFunc(data, {}, output) ?? {};
+
+            if (typeof outputStrings !== 'object') {
+              assert.fail(`'${trigger.id}' built-in response did not set outputStrings.`);
+              continue;
+            }
           }
         } else {
           if (trigger.outputStrings && typeof outputStrings !== 'object') {

